### DISCUSSION TITLE
Controls to integrate for pagination

### DISCUSSION THREAD

**Discussion by ApoorvaPetkar** [unoplatform/uno#9818] [Q&A]
Hi,

can someone please suggest a component for pagination purpose? Not able to find any in the winui3 desktop app and in the community toolkit I just found PipPager- https://docs.microsoft.com/en-us/windows/apps/design/controls/pipspager.

Is there any other component I can look into? Thanks in advance.

**Comment by jeromelaban** [comment]
Can you detail a bit more what you mean by pagination?

**Reply by ApoorvaPetkar** [reply]
@jeromelaban something like this to navigate across different pages. 

![image](https://user-images.githubusercontent.com/34087324/189902421-8f3088ac-4cac-43ca-835a-75d16e992b9b.png)

**Comment by MartinZikmund** [comment]
@ApoorvaPetkar You should be able to utilize `PagerControl` which is already ported in Uno Platform. You can follow the WinUI docs [here]( https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.pagercontrol?view=winui-2.8-prerelease)

**Reply by ApoorvaPetkar** [reply]
Hi @MartinZikmund  Thank you!

can you please provide any example link it would be great? 
We are using Windows community toolkit datagrid for which we want to apply pagination for.

**Comment by ApoorvaPetkar** [comment]
Hi @MartinZikmund,

The PagerControl you suggested just applies to winui2

![image](https://user-images.githubusercontent.com/34087324/190559012-e3e3408a-28b5-4440-b417-1303d6510dbf.png)

I'm using winui3, is there any alternative?

**Reply by MartinZikmund** [reply]
Unfortunately, the control is not yet available in Windows App SDK. I don't have an alternative right now (except for trying to copy the Uno Platform C# code of the control and using it on WinAppSDK too. Sorry for the late reply...

**Comment by sarosigy** [comment]
It seems that MS, starting with UWP, just moved from the concept of 'paging' to 'infinite list'. Like in Outlook inbox, which could have been implemented with paging, instead it just a list, where you can scroll down until ran out of items.

There is the `ISupportIncrementalLoading` interface which if you implement on a collection (along with `INotifyCollectionChanged`), the ListViewBase based controls (ListView, GridView) will try to load more and more items on scrolling.
(Which is problematic if you have a service which based on paging and expects a fixed page size: these controls very hectic on the expected number of items to be loaded.)

If you own the service (or the data source) and you can refactor (if needed) easily, I would move toward the `ISupportIncrementalLoading`. Ideally it is just a small adjustment on .Skip() and .Take() calls, and the result would me more 'modern Windows'-like.
If the service (or the data source) cannot be changed, or it would be a huge refactoring work, then it is not a big task to create a paging control, I remember, I created one within a day in the old WPF days.

So it seems, nowadays this is the direction.
